xdsl/dialects/builtin.py

```diff
@@ -21,6 +21,76 @@
     from xdsl.printer import Printer
 
 
+@dataclass
+class ArrayOfConstraint(AttrConstraint):
+    """
+    A constraint that enforces an ArrayData whose elements all satisfy
+    the elem_constr.
+    """
+    elem_constr: AttrConstraint
+
+    def __init__(self, constr: Attribute | Type[Attribute] | AttrConstraint):
+        self.elem_constr = attr_constr_coercion(constr)
+
+    def verify(self, attr: Attribute) -> None:
+        if not isinstance(attr, Data):
+            raise Exception(f"expected data ArrayData but got {attr}")
+        for e in cast(ArrayAttr[Attribute], attr).data:
+            self.elem_constr.verify(e)
+
+
+_ArrayAttrT = TypeVar("_ArrayAttrT", bound=Attribute, covariant=True)
+
+
+@irdl_attr_definition
+class ArrayAttr(GenericData[List[_ArrayAttrT]]):
+    name = "array"
+
+    @staticmethod
+    def parse_parameter(parser: BaseParser) -> List[_ArrayAttrT]:
```

Comment:
This function will fail, and also is never executed.
Can we just return an `NotImplementedError`, or an `assert`? These methods will never be called anyway since we have special support for them.

Reply:
Same for `print_parameter`

Reply:
Nevermind, this was from our old code, I'll fix it in another PR.